File: api/src/main/java/net/blueshell/api/service/ContributionPeriodService.java
package net.blueshell.api.service;

import net.blueshell.api.base.BaseModelService;
import net.blueshell.api.model.contribution.ContributionPeriod;
import net.blueshell.api.repository.ContributionPeriodRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.context.ApplicationEventPublisher;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

@Service
public class ContributionPeriodService extends BaseModelService<ContributionPeriod, ContributionPeriodRepository> {
    @Autowired
    public ContributionPeriodService(ContributionPeriodRepository repository, ApplicationEventPublisher events) {
        super(repository);
    }

    @Transactional(readOnly = true)
    public ContributionPeriod findLatest() {
        return repository.findCurrentOrLatestContributionPeriod();
    }
}

File: api/src/main/kotlin/net/blueshell/api/feature/contribution/service/ContributionPeriodService.kt
package net.blueshell.api.feature.contribution.service

import net.blueshell.api.feature.contribution.model.ContributionPeriod
import net.blueshell.api.feature.contribution.repository.ContributionPeriodRepository
import net.blueshell.api.shared.service.BaseModelService
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.context.ApplicationEventPublisher
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional

@Service
class ContributionPeriodService @Autowired constructor(
    repository: ContributionPeriodRepository,
    events: ApplicationEventPublisher
) : BaseModelService<ContributionPeriod, Long, ContributionPeriodRepository>(repository) {
    @Transactional(readOnly = true)
    fun findLatest(): ContributionPeriod {
        return repository!!.findCurrentOrLatestContributionPeriod()
    }

    @Transactional
    fun updateListId(periodId: Long, listId: Long) {
        val period = findById(periodId)
        period.listId = listId
        update(period)
    }
}
